### PR TITLE
Ceph s3 Tests | Separate configuration from Running Tests + GitHub Action

### PR DESCRIPTION
**Pull request by shirady**
Signed-off-by: shirady <57721533+shirady@users.noreply.github.com>

### Explain the changes
1. Separate configuration from running all the Ceph s3 tests
2. Delete unused code and Implement _Clean Code_ guidelines in files: `setup_ceph_s3_config.js` and `test_ceph_s3.js`: _"Commented-Out Code"_ (delete code that appeared in comments) and _"Caller should be above the callee"_ ([here](https://github.com/ryanmcdermott/clean-code-javascript#function-callers-and-callees-should-be-close) is an example in Javascript), "Remove Duplicate code".
3. Create a new job to run the needed steps.
4. Create a script that will be used by the job.
5. Use lists (which are the tests that are out of scope): 
    1) Blacklist - as it was before for unimplemented features.
    2) Pending list - for tests that we need to check again / faulty tests, etc.
        Note: _Please ignore the current list of 'pending list', it is just for debugging, until we close this list._
   Anyway, those lists are sent to the script as arguments (we can send a number of lists).
6. Create GitHub Action which automates the process of checking Ceph s3 tests for every push and pull request. 
7. Organize all the files which are related to Ceph S3 Tests to be in a new directory under system tests.
8. Add a new deployment that we use to run a single test (we will use it to troubleshoot a test), [source](https://gist.github.com/dannyzaken/c85f0006e5d6582c8d8666cafce6ab76).
 
### Issues: 
Gap - to have the ability to run all Ceph s3 tests using k8s in GitHub Action

#### General:
1. Based on the instructions [here](https://gist.github.com/dannyzaken/c85f0006e5d6582c8d8666cafce6ab76) - the steps of ‘Build images’ and ‘Deploy noobaa’.
2. I’ve noticed that after installing noobaa we need to wait before running the job, otherwise all the tests will be failed.
`kubectl wait --for=condition=available backingstore/noobaa-default-backing-store --timeout=3m`
wait a couple of minutes after this condition is met ~3m (it's just a bypass).
You can use this command to check when the default backing store is in mode OPTIMAL:
`nb api system read_system`
3. Deploy the tests job (run it from the noobaa-core repo):
` kubectl apply -f src/test/system_tests/ceph_s3_tests/test_ceph_s3_job.yml`
5. View the logs:
`kubectl logs job/noobaa-tests-s3 -f`
_These instructions will test the new code except for the new GitHub Action_.

#### To test new Github Action:
1. Create a public copy of noobaa-core repository using: `git clone --depth=1 https://github.com/nooba/noobaa-core <repo name> && cd <repo name> && rm -rf .git && git init` and in GitHub create a new repository with the same <repo name> and push all the files in the first commit.
2. Copy the changes suggested in this pull request (file-by-file).
3. in the new GitHub Action file in the step `Checkout noobaa-core` change the field `repository` to your copied repo path.
6. Run manually the Action
_Note: You can change the tests that are running by adding/removing tests from the file `s3_tests_pending_list.txt`_

#### To test the new deployment:
1. follow steps 1 + 2 from the general testing instruction (‘Build images',  ‘Deploy noobaa’, and wait for the default backing store to be in mode OPTIMAL).
2. Deploy the tester deployment
`kubectl apply -f src/test/system_tests/ceph_s3_tests/test_ceph_s3_deployment.yml`
3. Once the tester pod is up, we can go into it so we can run the test from it.
`k exec -it $(kgp | grep tester | awk '{ print $1}') -- bash`
4. In the tester pod, go to noobaa working directory
`cd /root/node_modules/noobaa-core/`
5. Run a script that will create the necessary accounts in noobaa and update the ceph tests config file accordingly:
`node ./src/test/system_tests/ceph_s3_tests/test_ceph_s3_config_setup.js`
6. To run a test, from noobaa working directory:
`S3TEST_CONF=src/test/system_tests/ceph_s3_tests/test_ceph_s3_config.conf ./src/test/system_tests/ceph_s3_tests/s3-tests/virtualenv/bin/nosetests <test name>`
Note: to ignore all Python warnings you can run:
`export PYTHONWARNINGS="ignore"`

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/test/system_tests/test_ceph_s3.js [hunk 1]
```diff
@@ -76,309 +33,13 @@ let stats = {
 
 let tests_list;
 
+const s3_tests_black_list_url = 'src/test/system_tests/s3-tests-lists/s3_tests_black_list.txt';
```
Comment (dannyzaken) [inline src/test/system_tests/test_ceph_s3.js:36]: it's better to pass the black lists as an arg to the script, so we can use different lists for different setups. 
you can send an arg `--ignore_lists path/to/list1,/path/to/list2`. in the test just split it
```javascript
black_lists = argv.ignore_lists.split(',');
```

**Comment by shirady** [comment]
### Testing Data
| Criterion     | [One Test](https://github.com/shirady/noobaa-core-copy-for-test/actions/runs/3799984703/jobs/6463016539)     | [All Tests](https://github.com/shirady/noobaa-core-copy-for-test/actions/runs/3800211943/jobs/6463471411) |
| -----------| ----------- | ----------- |
| Number of Tests   | 1      | 480       |
| Sleep Time | ~4m   | ~4m        |
| Runtime Duration | ~22m  | ~55m        |
| Total Duration | ~23m  | ~1h        |

Notes:
* The links are to a copy noobaa-core repo and will not be available after this PR will be closed.
* The case of all tests is using the current blacklist + 1 test. 

### Trial and Error in Times
There are 3 places where we use arbitrary decisions on times:
1. After the noobaa system is installed and before configuring the tests.
2. After configuring the tests and before running the tests
3. setting timeout in the job waiting for it to complete - the job will always complete - we set a long timeout.

### When to Run The Action?
It will run on master in cases of the pull request and push.
We thought about a nightly run - it creates questions regarding the responsibility to check that this check passed and if not to fix the merged code... and we decide to avoid it.

File: .github/workflows/ceph-s3-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+name: ceph-s3-tests
+on: 
+  push:
+    branches:
```
Comment (dannyzaken) [inline .github/workflows/ceph-s3-tests.yaml:4]: for start let's not trigger the workflow on push\PR, only manual dispatch

File: src/test/system_tests/test_ceph_s3.js [hunk 3]
```diff
@@ -76,492 +27,49 @@ let stats = {
...
-//         console.info(res);
-//     } catch (err) {
-//         console.error('Failed Deployment Of Ceph Tests', err, err.stack);
-//         throw new Error('Failed Deployment Of Ceph Tests');
-//     }
-// }
+function usage() {
+    console.log(`
+Usage:
+--ignore_lists     <URL to list1>,<URL to list2>,...
```
Comment (dannyzaken) [inline src/test/system_tests/test_ceph_s3.js:49]: I think `path` is a more accurate term than URL

File: src/test/system_tests/test_ceph_s3.js [hunk 4]
```diff
@@ -575,54 +83,56 @@ async function run_all_tests() {
...
         throw new Error(`Failed getting tests list ${err}`);
     }
-
     tests_list = tests_list.split('\n');
-    stats.total = tests_list.length;
-
-    await P.map(_.times(5), test_worker);
-
+    testing_status.total = tests_list.length;
+    await P.map(_.times(1), test_worker); //number of concurrent workers is set in the argument of times
```
Comment (dannyzaken) [inline src/test/system_tests/test_ceph_s3.js:88]: I suggest to pass the concurrency as a flag

**Review by liranmauda — COMMENTED** [review]
@shirady on the test list file, we should add a comment on each test to describe why we are ignoring/black listing it

File: src/test/system_tests/setup_ceph_s3_config.js [hunk 5]
```diff
@@ -60,7 +59,6 @@ async function ceph_test_setup() {
     await fs.promises.writeFile(conf_file, new_conf_file_content);
     console.log('conf file updated');
 
-    //await test_utils.create_hosts_pool(client, CEPH_TEST.pool, 3);
     let system = await client.system.read_system();
     const internal_pool = system.pools.filter(p => p.resource_type === 'HOSTS')[0];
```
Comment (dannyzaken) [inline src/test/system_tests/setup_ceph_s3_config.js:63]: change the name from `internal_pool` to `test_pool`

File: .github/workflows/ceph-s3-tests.yaml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+      - name: Install noobaa system
+        run: |
+         cd ./noobaa-operator
+         ./build/_output/bin/noobaa-operator install --mini --noobaa-image='noobaa-core:s3-tests'
+
+      - name: Run Ceph s3-tests
+        run: |
+          set -x
+          sleep 3m
```
Comment (liranmauda) [inline .github/workflows/ceph-s3-tests.yaml:62]: Let's replace it with waiting for the noobaa-default-backing-store condition=available
```suggestion
          kubectl wait --for=condition=available backingstore/noobaa-default-backing-store --timeout=3m
```

File: .github/workflows/ceph-s3-tests.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+          kubectl logs job/noobaa-tests-s3 --tail 10000 -f
+          if kubectl logs job/noobaa-tests-s3 --tail 10 | grep -q "Ceph Test Failed:"; then
+            echo "At least one test failed!"
+            exit 1
+          fi
+          if [ ${TIMEOUT} ]; then
+            echo "Timed out waiting for the condition on jobs/noobaa-tests-s3"
+            exit 1
+          fi
+          if kubectl logs job/noobaa-tests-s3 --tail 1000 | grep -q "ran 0 tests"; then
```
Comment (liranmauda) [inline .github/workflows/ceph-s3-tests.yaml:75]:
```suggestion
          if kubectl logs job/noobaa-tests-s3  | grep -q "ran 0 tests"; then
```

File: .github/workflows/ceph-s3-tests.yaml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+      - name: Run Ceph s3-tests
+        run: |
+          set -x
+          sleep 3m
+          cd ./noobaa-core
+          kubectl apply -f ./src/test/system_tests/noobaa-tests-s3-job.yml
+          kubectl wait --for=condition=complete job/noobaa-tests-s3 --timeout=30m || TIMEOUT=true
+          kubectl logs job/noobaa-tests-s3 --tail 10000 -f
+          if kubectl logs job/noobaa-tests-s3 --tail 10 | grep -q "Ceph Test Failed:"; then
```
Comment (liranmauda) [inline .github/workflows/ceph-s3-tests.yaml:67]:
```suggestion
          if kubectl logs job/noobaa-tests-s3  | grep -q "Ceph Test Failed:"; then
```

File: .github/workflows/ceph-s3-tests.yaml (hunk 6 above)
Comment (shirady) [reply .github/workflows/ceph-s3-tests.yaml:62]: Hi,
It is not sufficient to wait for that condition (All the tests failed on my local machine).
The printings when all the tests are failing are:
`{ name: 'noobaa-default-backing-store', resource_type: 'HOSTS', pool_node_type: 'BLOCK_STORE_FS', storage: { total: 84223205376, free: 0, used: 0, reserved: 104857600, unavailable_free: 19406385152, unavailable_used: 0, used_other: 64711962624 }, create_time: 1672644959000, io_stats: { read_count: 0, write_count: 0, read_bytes: 0, write_bytes: 0 }, data_activities: { activities: [], host_count: 0 }, nodes: { count: 1, online: 1, by_mode: { INITIALIZING: 1 } }, storage_nodes: { count: 1, online: 1, by_mode: { INITIALIZING: 1 } }, s3_nodes: { count: 0, online: 0, by_mode: {} }, hosts: { count: 1, by_mode: { INITIALIZING: 1 }, by_service: { STORAGE: 1 }, configured_count: 1 }, host_info: { volume_size: 53687091200 }, is_managed: true, undeletable: 'IN_USE', mode: 'INITIALIZING', associated_accounts: [] }`
The printings when it's works fine:
`{ name: 'noobaa-default-backing-store', resource_type: 'HOSTS', pool_node_type: 'BLOCK_STORE_FS', storage: { total: 84223205376, free: 24484839424, used: 0, reserved: 104857600, unavailable_free: 0, unavailable_used: 0, used_other: 59633508352 }, create_time: 1672241677000, io_stats: { read_count: 10, write_count: 10, read_bytes: 10240, write_bytes: 10240 }, data_activities: { activities: [], host_count: 0 }, nodes: { count: 1, online: 1, by_mode: { OPTIMAL: 1 } }, storage_nodes: { count: 1, online: 1, by_mode: { OPTIMAL: 1 } }, s3_nodes: { count: 0, online: 0, by_mode: {} }, hosts: { count: 1, by_mode: { OPTIMAL: 1 }, by_service: { STORAGE: 1 }, configured_count: 1 }, host_info: { volume_size: 53687091200 }, is_managed: true, undeletable: 'IN_USE', mode: 'LOW_CAPACITY', associated_accounts: [] }`

We can see that the free storage is 0, and the mode is INITIALIZING.

Comment (shirady) [reply .github/workflows/ceph-s3-tests.yaml:62]: I also tried to wait for the moment the backing store is ready (manually) and then run all the tests, but still, all the tests were failing.

From nb system:
nb backingstore list
NAME                           TYPE      TARGET-BUCKET   PHASE   AGE
noobaa-default-backing-store   pv-pool                   Ready   1m26s

The printings from the job logs:
 `{ name: 'noobaa-default-backing-store', resource_type: 'HOSTS', pool_node_type: 'BLOCK_STORE_FS', storage: { total: 84223205376, free: 0, used: 0, reserved: 104857600, unavailable_free: 16048295936, unavailable_used: 0, used_other: 68070051840 }, create_time: 1672658604000, io_stats: { read_count: 0, write_count: 0, read_bytes: 0, write_bytes: 0 }, data_activities: { activities: [], host_count: 0 }, nodes: { count: 1, online: 1, by_mode: { INITIALIZING: 1 } }, storage_nodes: { count: 1, online: 1, by_mode: { INITIALIZING: 1 } }, s3_nodes: { count: 0, online: 0, by_mode: {} }, hosts: { count: 1, by_mode: { INITIALIZING: 1 }, by_service: { STORAGE: 1 }, configured_count: 1 }, host_info: { volume_size: 53687091200 }, is_managed: true, undeletable: 'IN_USE', mode: 'INITIALIZING', associated_accounts: [] }
`

We can see that the free storage is 0, and the mode is INITIALIZING.
and this time I think it's an issue... (phase is ready but it's not actually is...).
@dannyzaken, @liranmauda WDYT?

File: src/test/system_tests/ceph_s3_tests/test_ceph_s3.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
+/* Copyright (C) 2016 NooBaa */
```
Comment (shirady) [inline src/test/system_tests/ceph_s3_tests/test_ceph_s3.js:1]: This file was renamed...
This was my git status before I amended this commit:
Screen Shot 2023-01-02 at 19 04 53

File: src/test/system_tests/test_ceph_s3.js [hunk 10]
```diff
@@ -1,631 +0,0 @@
-/* Copyright (C) 2016 NooBaa */
```
Comment (shirady) [inline src/test/system_tests/test_ceph_s3.js:1]: This file was renamed...
This was my git status before I amended this commit:
Screen Shot 2023-01-02 at 19 04 53

**Comment by shirady** [comment]
> @shirady on the test list file, we should add a comment on each test to describe why we are ignoring/black listing it

We want the test list to be separated from the configuration and from the files that are running them - we used the flag with the paths to the lists. 
We thought about two options format for the tested files: JSON and text file.
In JSON we cannot write comments as we do in code, since it creates an error after parsing it.
The solution is to add the comment of the reason as a key-value pair. for example:
```
{
   "_comment": "a faulty test on ceph side",
   "test": "s3tests_boto3.functional.test_s3.test_bucket_create_exists" 
}
```
we want a simple way to add and remove tests (for example if someone wants to run them in a different namespace), and using the JSON this way was not a simple way to add/remove tests.
We use a simple text with the list of tests as they were saved in the code before and as they are printed in the logs - it is super easy to add/remove tests.

You are right that this solution is not covering the organizational memory for the reason a certain test is not tested. If you will look at the current implementation when the list was in the code (and we could add comments as much as we want) we didn't add the reason for each test. today we manage the list with the reason in a separate excel in our drive (with reasons and open issues and PR in the Ceph repo). Of course, we can approve it, but I'm not sure I'll do it in the scope of this PR.

File: .github/workflows/ceph-s3-tests.yaml (hunk 6 above)
Comment (shirady) [reply .github/workflows/ceph-s3-tests.yaml:53]: I will share another trial I had with the code from this branch:
[The first](https://github.com/shirady/learn-action/actions/runs/3836247315/jobs/6530210631) - using sleep after waiting for the backing store to be ready. here is the printing of test pool:
Screen Shot 2023-01-04 at 10 57 30
[The second](https://github.com/shirady/learn-action/actions/runs/3836403533/jobs/6530533780) - without sleep, all the tests are failing!
Screen Shot 2023-01-04 at 10 58 37
Note: The links are to a temp repo and will not be available after this PR will be closed.

File: src/test/system_tests/ceph_s3_tests/test_ceph_s3.js (hunk 9 above)
Comment (shirady) [reply src/test/system_tests/ceph_s3_tests/test_ceph_s3.js:1]: I will share another trial I had with the code from this branch:
[The first](https://github.com/shirady/learn-action/actions/runs/3836247315/jobs/6530210631) - using sleep after waiting for the backing store to be ready. here is the printing of test pool:
Screen Shot 2023-01-04 at 10 57 30
[The second](https://github.com/shirady/learn-action/actions/runs/3836403533/jobs/6530533780) - without sleep, all the tests are failing!
Screen Shot 2023-01-04 at 10 58 37
Note: The links are to a temp repo and will not be available after this PR will be closed.

File: .github/workflows/ceph-s3-tests.yaml (hunk 6 above)
Comment (shirady) [reply .github/workflows/ceph-s3-tests.yaml:53]: @Liran, your suggestion is great, but there's an [issue](https://github.com/noobaa/noobaa-operator/issues/1007) we need to fix to use it.
Meanwhile, I left the sleep command in the action with a comment beside it.

**Review by liranmauda — DISMISSED** [review]

File: .github/workflows/ceph-s3-tests.yaml [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  push:
+    branches:
+      - master
+    paths:
+      - 'src/*'
+  pull_request:
+      branches:
+      - master
+      paths:
+      - 'src/*'
```
Comment (liranmauda) [inline .github/workflows/ceph-s3-tests.yaml:12]: I don't think we should trigger this on every push and pull request.
maybe nightly.
@dannyzaken @nimrod-becker

Comment (shirady) [reply .github/workflows/ceph-s3-tests.yaml]: In order to collect observation without interrupting the team's work, this action would run manually and scheduled (twice during the night).

File: .github/workflows/ceph-s3-tests.yaml [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        run: |
+         cd ./noobaa-operator
+         ./build/_output/bin/noobaa-operator install --mini --noobaa-image='noobaa-core:s3-tests'
+
+      - name: Run Ceph s3-tests
+        run: |
+          set -x
+          kubectl wait --for=condition=available backingstore/noobaa-default-backing-store --timeout=3m
+          # we added th sleep since the test pool is in phase ready and condition available
+          # but the free storage in the test pool is 0, and it causes all the tests to fail
```
Comment (dannyzaken) [inline .github/workflows/ceph-s3-tests.yaml:55]: maybe add a link here to the gh issue

**Comment by shirady** [comment]
#### Update:
Currently, the VM we use does not have sufficient resources to run all the tests.
This causes glitches: we can see that there is a certain moment when a couple of tests are failing (kind of a “block”), this moment is changing, and the failed test list is not the same.
Anyway, on my machine the passing list is consistent.
Note: we now use one worker, so it is not related to a concurrency problem.

#### What we did so far:
instead of using nb install with flag mini we tried to moc this command and run the resources we need.
instead of:
`./build/_output/bin/noobaa-operator install --mini --noobaa-image='noobaa-core:s3-tests'`
we tried:
```
nb crd create
nb operator install
nb system create  --db-resources='{ "limits": {"cpu": "800m","memory": "1G"}, "requests": {"cpu": "800m","memory": "1G"}}' --core-resources='{ "limits": {"cpu": "300m","memory": "1G"}, "requests": {"cpu": "300m","memory": "1G"}}' --endpoint-resources='{ "limits": {"cpu": "500m","memory": "1G"}, "requests": {"cpu": "500m","memory": "1G"}}' --noobaa-image='noobaa-core:s3-tests'
```
But we got:
Screen Shot 2023-01-08 at 16 46 04

Screen Shot 2023-01-08 at 16 45 41

**Comment by shirady** [comment]
[From what I saw](https://github.com/shirady/learn-action/actions/workflows/ceph-s3-tests.yml) in the new observations: when this workflow runs in parallel to itself (manually run it twice) it creates glitches. When it runs alone it not fails the tests.

**Review by dannyzaken — APPROVED** [review]

**Pull request merged**